haystack/document_stores/elasticsearch8.py

```diff
@@ -0,0 +1,335 @@
+                    request_timeout=timeout,
+                    node_class=node_class,
+                )
+        else:
+            # there is no authentication for this elasticsearch instance
+            if ca_certs is not None:
+                client = Elasticsearch(
+                    hosts=hosts,
+                    ca_certs=ca_certs,
+                    verify_certs=verify_certs,
+                    request_timeout=timeout,
+                    node_class=node_class,
+                )
+            else:
+                client = Elasticsearch(
+                    hosts=hosts,
+                    basic_auth=(username, password),
+                    verify_certs=verify_certs,
+                    request_timeout=timeout,
+                    node_class=node_class,
+                )
+
+        # Test connection
+        try:
+            # ping uses a HEAD request on the root URI. In some cases, the user might not have permissions for that,
+            # resulting in a HTTP Forbidden 403 response.
+            if username in ["", "elastic"]:
+                status = client.ping()
+                if not status:
+                    raise ConnectionError(
+                        f"Initial connection to Elasticsearch failed. Make sure you run an Elasticsearch instance "
+                        f"at `{hosts}` and that it has finished the initial ramp up (can take > 30s). Also, make sure "
+                        f"you are using the correct credentials if you are using a secured Elasticsearch instance."
+                    )
+        except Exception:
+            raise ConnectionError(
+                f"Initial connection to Elasticsearch failed. Make sure you run an Elasticsearch instance at `{hosts}` "
+                f"and that it has finished the initial ramp up (can take > 30s). Also, make sure you are using the "
+                f"correct credentials if you are using a secured Elasticsearch instance."
+            )
+        return client
+
+    @staticmethod
+    def _prepare_hosts(host: Union[str, List[str]], port: Union[int, List[int]], scheme: str):
+        """
+        Create a list of host(s), port(s) and scheme to allow direct client connections to multiple nodes,
+        in the format expected by the client.
+        """
+        if isinstance(host, list):
+            if isinstance(port, list):
+                if not len(port) == len(host):
+                    raise ValueError("Length of list `host` must match length of list `port`")
+                hosts = [{"host": h, "port": p, "scheme": scheme} for h, p in zip(host, port)]
+            else:
+                hosts = [{"host": h, "port": port, "scheme": scheme} for h in host]
+        else:
+            hosts = [{"host": host, "port": port, "scheme": scheme}]
+        return hosts
+
+    def _do_bulk(
+        self, actions: Union[List[Dict]], client: "Elasticsearch", refresh: str, headers: Optional[Dict]  # type: ignore[override]
```

Comment:
What's the deal with the `# type: ignore[override]`? If it makes sense let's document it.